### ISSUE TITLE
Use hard links when sandboxing on Windows when possible

### ISSUE THREAD

**Issue by snowleopard** [ocaml/dune#4362]
In #4360 we added basic support for sandboxing actions with hard links. Currently, on Windows the implementation falls back to copying because it's unclear whether hard links work consistently across different Windows configurations.

Instead of always falling back to copying, we could detect if hard linking works by trying it, and if it does, use it. We already do something like this in the Dune cache implementation, so we can share some code.

**Comment by rgrinberg** [comment]
Yes, if it works for you with `--sandbox=hardlink` and the performance is acceptable, we'll be much more comfortable pushing ahead with this.

**Comment by nojb** [comment]
> I will give it a try and report back.

A full rebuild of the LexiFi codebase under Windows takes ~4m43s without the cache and ~5m4s using the cache with hardlinks (~10% slower). Using the cache with copying takes ~5m10s, so not much difference between hardlinks and copying.

**Comment by rgrinberg** [comment]
@jonahbeckford could you give us a datapoint here with diskuv?

**Comment by jonahbeckford** [comment]
(I'm unfamiliar with Dune caching myself.)

Is it just a matrix of `DUNE_SANDBOX={none,hardlink}` on a large build? Or is there something else that needs to be turned on?

**Comment by rgrinberg** [comment]
> (I'm unfamiliar with Dune caching myself.)

Note that we aren't touching caching here. Sandboxing just makes sure that rules cannot observe files on which they haven't declared a dependency on.

Although it would be awesome if you could make sure that caching works on Windows as well.

> Is it just a matrix of DUNE_SANDBOX={none,hardlink} on a large build? Or is there something else that needs to be turned on?

Yup, that's all it would take. Note that you can also pass `--sandbox hardlink`.

**Comment by jonahbeckford** [comment]
@rgrinberg : I get the following when I set `DUNE_SANDBOX=hardlink` in a [Windows runner on GitLab CI/CD](https://docs.gitlab.com/ee/ci/runners/saas/windows_saas_runner.html):

```
[dkml-apps-4624-f0e89f.out] Internal error, please report upstream including the contents of _build/log.
[dkml-apps-4624-f0e89f.out] Description:
[dkml-apps-4624-f0e89f.out]   ("Don't have hardlinks on win32, but [hardlink] sandboxing mode was selected. To use emulation via copy, the [copy] sandboxing mode should be selected.",
[dkml-apps-4624-f0e89f.out]   {})
[dkml-apps-4624-f0e89f.out] Raised at Stdune__code_error.raise in file "otherlibs/stdune\\code_error.ml",
[dkml-apps-4624-f0e89f.out]   line 11, characters 30-62
[dkml-apps-4624-f0e89f.out] Called from Dune_engine__sandbox.link_function.win32_error in file
[dkml-apps-4624-f0e89f.out]   "src/dune_engine\\sandbox.ml", line 47, characters 4-237
[dkml-apps-4624-f0e89f.out] Called from Dune_engine__sandbox.link_deps in file
[dkml-apps-4624-f0e89f.out]   "src/dune_engine\\sandbox.ml", line 73, characters 28-49
[dkml-apps-4624-f0e89f.out] Called from Dune_engine__sandbox.create in file
[dkml-apps-4624-f0e89f.out]   "src/dune_engine\\sandbox.ml", line 126, characters 2-25
[dkml-apps-4624-f0e89f.out] Called from Dune_engine__build_system.Exported.execute_action_for_rule in
[dkml-apps-4624-f0e89f.out]   file "src/dune_engine\\build_system.ml", line 416, characters 10-237
[dkml-apps-4624-f0e89f.out] Called from Fiber__scheduler.exec in file "src/fiber\\scheduler.ml", line 73,
[dkml-apps-4624-f0e89f.out]   characters 8-11
[dkml-apps-4624-f0e89f.out] Re-raised at Stdune__exn.raise_with_backtrace in file
[dkml-apps-4624-f0e89f.out]   "otherlibs/stdune\\exn.ml" (inlined), line 36, characters 27-56
[dkml-apps-4624-f0e89f.out] Called from Stdune__exn_with_backtrace.reraise in file
[dkml-apps-4624-f0e89f.out]   "otherlibs/stdune\\exn_with_backtrace.ml", line 18, characters 33-71
[dkml-apps-4624-f0e89f.out] Called from Fiber__scheduler.exec in file "src/fiber\\scheduler.ml", line 73,
[dkml-apps-4624-f0e89f.out]   characters 8-11
```

FYI: I've had poor experience with symlinks on Windows (something I've never seen a need to use). Another is the bug I created at https://github.com/ocaml/opam/issues/5368 .

**Comment by punchagan** [comment]
Closed in #13987

**Comment by Alizter** [comment]
@punchagan We are currently not hard linking by default on Windows, which are just allowing it to be configured. Therefore I am keeping this issue open.

**Comment by Alizter** [comment]
@jonahbeckford In 3.23 I've allowed using hardlinking for sandboxing on Windows, so you shouldn't have that error you ran into before. Would you be able to give it another shot?

**Comment by jonahbeckford** [comment]
Oh my. This is a quite old issue, and I don't build and don't have the same infrastructure to recreate the exact command + behavior from 4 years ago. If I see an opportunity to test something similar in gitlab I will, but please don't gate on me.

Thanks though!
